src/main/java/com/ibm/cloud/sdk/core/security/IamToken.java

```diff
@@ -51,28 +52,51 @@ public Long getExpiration() {
     return expiration;
   }
 
+  IamToken() {
+    if (getExpiresIn() != null && getExpiration() != null) {
+      Double fractionOfTimeToLive = 0.8;
+      Long timeToLive = getExpiresIn();
+      Long expirationTime = getExpiration();
+      this.refreshTime = expirationTime - (long) (timeToLive * (1.0 - fractionOfTimeToLive));
+    }
+  }
+
   /**
-   * Check if currently stored access token is valid.
+   * Returns true iff currently stored access token should be refreshed.
    *
-   * Using a buffer to prevent the edge case of the
-   * token expiring before the request could be made.
+   * This method uses a buffer to prevent the edge case of the
+   * token expiring before the request could be made, so this method will return true
+   * if the current time is within that buffer.
    *
    * The buffer will be a fraction of the total TTL. Using 80%.
    *
-   * @return true iff the current access token is valid and not expired
+   * This method also updates the expiration time if it determines the token needs refreshed to prevent other
+   * threads from making multiple refresh calls.
+   *
+   * @return true iff the current access token is invalid or past the expiration buffer
    */
   @Override
-  public boolean isTokenValid() {
-    if (getAccessToken() == null || getExpiresIn() == null || getExpiration() == null) {
-      return false;
+  public synchronized boolean needsRefresh() {
+    long currentTime = System.currentTimeMillis() / 1000;
+
+    if (refreshTime == null || currentTime > refreshTime) {
```

Comment:
I think we need to be clear about what `refreshTime == null` means.  Does it mean "this token needs to be refreshed immediately" or does it mean "never refresh this token"?  I would have thought it was the latter, but here it seems to be taken to mean the former.

Reply:
Looking at it again maybe it should mean the former.

Originally I was thinking that if the `refreshTime` wasn't set, we were parsing some sort of invalid IAM response that was missing fields. It does seem possible though that there just wouldn't be `expires_in` or `expiration` fields in the response to indicate that it doesn't expire at all.